### PR TITLE
executor: use `WithRecovery` in IndexMergeReaderExecutor instead of `recover`

### PR DESCRIPTION
**Pull request by SeaRise**
### What problem does this PR solve? 
Fix issue #14335

### What is changed and how it works?
use `WithRecovery` in IndexMergeReaderExecutor instead of `recover`

### Check List

Tests 
 - Unit test
 - Integration test

Code changes
 - Has interface methods change

### CODE REVIEW

File: executor/index_merge_reader.go [hunk 1]
```diff
@@ -622,18 +642,9 @@ type indexMergeTableScanWorker struct {
 	memTracker *memory.Tracker
 }
 
-func (w *indexMergeTableScanWorker) pickAndExecTask(ctx context.Context) {
-	var task *lookupTableTask
+func (w *indexMergeTableScanWorker) pickAndExecTask(ctx context.Context) (task *lookupTableTask) {
+	//var task *lookupTableTask
```
Comment (zz-jason) [inline executor/index_merge_reader.go]: can we remove this line totally?

File: executor/index_merge_reader.go [hunk 2]
```diff
@@ -320,6 +308,28 @@ func (w *partialTableWorker) fetchHandles(ctx context.Context, fetchCh chan<- *l
...
 
+func (w *partialTableWorker) handleHandlesFetcherPanic(ctx context.Context, resultCh chan<- *lookupTableTask, err error) func(r interface{}) {
+	return func(r interface{}) {
+		if r == nil {
+			return
+		}
+
+		buf := make([]byte, 4096)
+		stackSize := runtime.Stack(buf, false)
+		buf = buf[:stackSize]
```
Comment (zz-jason) [inline executor/index_merge_reader.go]: no need to record the call stack in this function, `WithRecovery()` will handle this. See the implementation of `WithRecovery()` for details.

File: executor/index_merge_reader.go [hunk 3]
```diff
@@ -320,6 +307,25 @@ func (w *partialTableWorker) fetchHandles(ctx context.Context, fetchCh chan<- *l
 	}
 }
 
+func (w *partialTableWorker) handleHandlesFetcherPanic(ctx context.Context, resultCh chan<- *lookupTableTask, err error) func(r interface{}) {
+	return func(r interface{}) {
+		if r == nil {
+			return
+		}
+
+		logutil.Logger(ctx).Error("partialTableWorker in IndexMergeReaderExecutor panicked")
```
Comment (zz-jason) [inline executor/index_merge_reader.go]: how about:
```go
err4Panic := errors.Errorf("panic in IndexMergeReaderExecutor partialTableWorker: %v", r)
logutil.Logger(ctx).Error(err4Panic.Error())
doneCh <- err4Panic
```

File: executor/index_merge_reader.go [hunk 4]
```diff
@@ -188,7 +190,8 @@ func (e *IndexMergeReaderExecutor) startPartialIndexWorker(ctx context.Context,
 
 	go func() {
 		ctx1, cancel := context.WithCancel(ctx)
-		_, err := worker.fetchHandles(ctx1, result, fetchCh, e.resultCh, e.finished)
+		var err error
+		util.WithRecovery(func() { _, err = worker.fetchHandles(ctx1, result, fetchCh, e.resultCh, e.finished) }, worker.handleHandlesFetcherPanic(ctx, e.resultCh, err))
```
Comment (zz-jason) [inline executor/index_merge_reader.go]: how about:
```go
util.WithRecovery(
	func() { _, err = worker.fetchHandles(ctx1, result, fetchCh, e.resultCh, e.finished) },
	worker.handleHandlesFetcherPanic(ctx, e.resultCh, err),
)
```

File: executor/index_merge_reader.go [hunk 5]
```diff
@@ -237,7 +240,8 @@ func (e *IndexMergeReaderExecutor) startPartialTableWorker(ctx context.Context,
 	}
 	go func() {
 		ctx1, cancel := context.WithCancel(ctx)
-		_, err := worker.fetchHandles(ctx1, fetchCh, e.resultCh, e.finished)
+		var err error
+		util.WithRecovery(func() { _, err = worker.fetchHandles(ctx1, fetchCh, e.resultCh, e.finished) }, worker.handleHandlesFetcherPanic(ctx, e.resultCh, err))
```
Comment (zz-jason) [inline executor/index_merge_reader.go]: how about:
```go
util.WithRecovery(
	func() { _, err = worker.fetchHandles(ctx1, fetchCh, e.resultCh, e.finished) },
	worker.handleHandlesFetcherPanic(ctx, e.resultCh, err),
)
```

File: executor/index_merge_reader.go [hunk 6]
```diff
@@ -320,6 +307,25 @@ func (w *partialTableWorker) fetchHandles(ctx context.Context, fetchCh chan<- *l
...
+		}
+
+		logutil.Logger(ctx).Error("partialTableWorker in IndexMergeReaderExecutor panicked")
+		err4Panic := errors.Errorf("%v", r)
+		doneCh := make(chan error, 1)
+		doneCh <- err4Panic
+		resultCh <- &lookupTableTask{
+			doneCh: doneCh,
+		}
+		if err != nil {
```
Comment (zz-jason) [inline executor/index_merge_reader.go]: should it be `if err == nil`?

Comment (SeaRise) [reply executor/index_merge_reader.go]: I made a mistake...
`if (err != nil)  err = errors.Trace(err4Panic)` will add `err4Panic` in `error`.
and `error` with `err4Panic` will be printed by `logutil.Logger(ctx).Error("close Select result failed:", zap.Error(err))`.

so ... `if err != nil {` in there is senseless。
may be I should return error in `worker.handleHandlesFetcherPanic(ctx, e.resultCh, err)`

Comment (SeaRise) [reply executor/index_merge_reader.go]: `if err != nil { err = errors.Trace(err4Panic) }`
can I remove this code in there?
It seems to be meaningless。

Comment (zz-jason) [reply executor/index_merge_reader.go]: OK

File: executor/index_merge_reader.go [hunk 7]
```diff
@@ -320,6 +316,22 @@ func (w *partialTableWorker) fetchHandles(ctx context.Context, fetchCh chan<- *l
 	}
 }
 
+func (w *partialTableWorker) handleHandlesFetcherPanic(ctx context.Context, resultCh chan<- *lookupTableTask) func(r interface{}) {
```
Comment (XuHuaiyu) [inline executor/index_merge_reader.go]: Actually, we do not need to define `handleHandlesFetcherPanic` as the member functions for  `partialTableWorker` and `partialIndexWorker` individually ?

File: executor/index_merge_reader.go [hunk 8]
```diff
@@ -648,6 +654,18 @@ func (w *indexMergeTableScanWorker) pickAndExecTask(ctx context.Context) {
 	}
 }
 
+func (w *indexMergeTableScanWorker) handlePickAndExecTaskPanic(ctx context.Context, task *lookupTableTask) func(r interface{}) {
```
Comment (XuHuaiyu) [inline executor/index_merge_reader.go]: Can we pass the `task.doneCh` instead of `task` here?

**Comment by XuHuaiyu** [comment]
@SeaRise CI failed because of this panic, can you check it ?

```
[2020-02-12T14:42:36.796Z] panic: runtime error: invalid memory address or nil pointer dereference
[2020-02-12T14:42:36.796Z] [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x21d7236]
[2020-02-12T14:42:36.796Z] 
[2020-02-12T14:42:36.796Z] goroutine 47679 [running]:
[2020-02-12T14:42:36.796Z] github.com/pingcap/tidb/executor.(*IndexMergeReaderExecutor).startIndexMergeTableScanWorker.func2(0xc004fae180, 0x2c9b280, 0xc004fae1c0, 0xc004f66510, 0xc010824480)
[2020-02-12T14:42:36.796Z] 	/home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/executor/index_merge_reader.go:375 +0x36
[2020-02-12T14:42:36.796Z] created by github.com/pingcap/tidb/executor.(*IndexMergeReaderExecutor).startIndexMergeTableScanWorker
[2020-02-12T14:42:36.796Z] 	/home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/executor/index_merge_reader.go:371 +0xff
[2020-02-12T14:42:36.796Z] FAIL	github.com/pingcap/tidb/executor	5.955s
```

File: executor/index_merge_reader.go (hunk 8 above)
Comment (SeaRise) [reply executor/index_merge_reader.go]: task had not init and task.doneCh is nil.

**Pull request merged**